docs/syntax.md

```diff
@@ -0,0 +1,255 @@
+`space`s or line comments are ignored except as it separates tokens that
+    would otherwise combine into a single token.
+
+### Line comments
+A *line comment* starts with the terminal `'--'` followed by zero or more printable Unicode characters stopping at the first end of line (`'\n'` or `'\r\n'`).
+
+## Syntax of LambdaBuffers files
+A LambdaBuffers file defines a module that is a collection of data types, classes, instance clauses, and derive clauses.
+
+The overall layout of a LambdaBuffers file is:
+
+```text
+module -> 'module' longmodulename { import } { statement }
+```
+
+The file must specify the module's `longmodulename` where its `modulename` must match the file's name not including the `.lbf` extension.
+After, the file may contain a sequence of `import`s followed by a sequence of `statement`s.
+
+### Import
+Imports bring *entities* (types and classes) of other modules into scope.
+
+```text
+import     -> 'import' [ 'qualified' ] longmodulename [ 'as' longmodulename ] [ importspec ]
+importspec -> '(' [ { typename ',' } typename [','] ] ')'
+```
+
+If `importspec` is omitted, then all entities specified in the module are imported; otherwise only the specified entities are imported.
+
+### Statement
+
+Statements define types, classes, instance clauses, and derive clauses.
+
+```text
+statement -> typedef
+           | classdef
+           | instanceclause
+           | deriveclause
+```
+
+#### Type definitions
+Types may be either sum types, product types, record types, or opaque types.
+
+```text
+typedef -> prodtypedef | sumtypedef |  recordtypedef | opaquetypedef
+```
+
+##### Product type definition
+A product type definition defines a new product type.
+
+```text
+prodtypedef -> 'prod' typename { varname } '=' prod
+prod        -> { typeexp }
+typeexp     -> varname
+             | longtypename
+             | '(' prod ')'
+```
+
+Product type definitions instruct the code generator to generate a product type for the target language.
+
+##### Sum type definition
+A sum type definition defines a new sum type.
+
+```text
+sumtypedef     -> 'sum' typename { varname } '=' sum
+sum            -> sumconstructor { '|' sumconstructor }
+sumconstructor -> typename prod
+```
+
+Sum type definitions instruct the code generator to generate a sum type for the target language.
+
+##### Record type definition
+A record type definition defines a new record type.
+
+```text
+recordtypedef -> 'record' typename { varname } '=' record
+record        -> '{' [ field { ',' field  } ] '}'
+field         -> fieldname ':' prod
+````
+
+Record type definitions instruct the code generator to generate a record type for the target language.
+
+##### Opaque type definition
+An opaque type definition defines a new opaque type.
+
+```text
+opaquetypedef -> 'opaque' typename { varname }
+```
+
+Opaque type definitions do not instruct the code generator to generate code, and an opaque type must be instead implemented in the target language.
```

Comment:
Hmm yes, we do generate type aliases to allow users to just use LB generate modules.

So if Foo.lbf

```
module Foo

opaque Int
```

and a configuration

```
{
  "Foo.Int": ["prelude", "Prelude.Int"]
}
```

we generate a LambdaBuffers/Foo.hs

```
module LambdaBuffers.Foo where
import Prelude

type Foo = Prelude.Int
```

Reply:
So maybe just saying smtn like; Opaque type definitions must map to existing types in a target language and it's up to the Codegen module to determine how that's exactly done.